File: src/test/java/org/vw/domain/CleaningRobotTest.java
package org.vw.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class CleaningRobotTest {
    private Grid grid;
    private CleaningRobot robot;

    @BeforeEach
    void setUp() {
        grid = new Grid(5, 5);
        Position startPosition = new Position(grid, 2, 2, Position.Direction.N);
        robot = new CleaningRobot(startPosition, new char[]{'M', 'L', 'M', 'R', 'M'});
    }

    @Test
    void testExecuteCommand_MoveForward() {
        robot.executeCommand('M');
        assertEquals(2, robot.getPosition().getX());
        assertEquals(3, robot.getPosition().getY());
        assertEquals(Position.Direction.N, robot.getPosition().getDirection());
    }

    @Test
    void testExecuteCommand_RotateLeft() {
        robot.executeCommand('L');
        assertEquals(Position.Direction.W, robot.getPosition().getDirection());
    }

    @Test
    void testExecuteCommand_RotateRight() {
        robot.executeCommand('R');
        assertEquals(Position.Direction.E, robot.getPosition().getDirection());
    }

    @Test
    void testExecuteCommand_EastBlocked() {
        robot.getPosition().setX(grid.getMaxX());
        robot.getPosition().setDirection(Position.Direction.E);
        robot.executeCommand('M');
        assertEquals(grid.getMaxX(), robot.getPosition().getX());
    }

    @Test
    void testExecuteCommand_NorthBlocked() {
        robot.getPosition().setY(grid.getMaxY());
        robot.getPosition().setDirection(Position.Direction.N);
        robot.executeCommand('M');
        assertEquals(grid.getMaxY(), robot.getPosition().getY());
    }

    @Test
    void testExecuteCommand_SouthBlocked() {
        robot.getPosition().setY(0);
        robot.getPosition().setDirection(Position.Direction.S);
        robot.executeCommand('M');
        assertEquals(0, robot.getPosition().getY());
    }

    @Test
    void testExecuteCommand_WestBlocked() {
        robot.getPosition().setX(0);
        robot.getPosition().setDirection(Position.Direction.W);
        robot.executeCommand('M');
        assertEquals(0, robot.getPosition().getX());
    }
}

File: src/test/java/org/vw/domain/CleaningRobotTest.kt
package org.vw.domain

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class CleaningRobotTest {
    private var grid: Grid? = null
    private var robot: CleaningRobot? = null

    @BeforeEach
    fun setUp() {
        grid = Grid(5, 5)
        val startPosition = Position(grid!!, 2, 2, Position.Direction.N)
        robot = CleaningRobot(startPosition, listOf('M', 'L', 'M', 'R', 'M'))
    }

    @Test
    fun testExecuteCommand_MoveForward() {
        robot!!.executeCommand('M')
        Assertions.assertEquals(2, robot!!.position.x)
        Assertions.assertEquals(3, robot!!.position.y)
        Assertions.assertEquals(Position.Direction.N, robot!!.position.direction)
    }

    @Test
    fun testExecuteCommand_RotateLeft() {
        robot!!.executeCommand('L')
        Assertions.assertEquals(Position.Direction.W, robot!!.position.direction)
    }

    @Test
    fun testExecuteCommand_RotateRight() {
        robot!!.executeCommand('R')
        Assertions.assertEquals(Position.Direction.E, robot!!.position.direction)
    }

    @Test
    fun testExecuteCommand_EastBlocked() {
        robot!!.position.x = grid!!.maxX
        robot!!.position.direction = Position.Direction.E
        robot!!.executeCommand('M')
        Assertions.assertEquals(grid!!.maxX, robot!!.position.x)
    }

    @Test
    fun testExecuteCommand_NorthBlocked() {
        robot!!.position.y = grid!!.maxY
        robot!!.position.direction = Position.Direction.N
        robot!!.executeCommand('M')
        Assertions.assertEquals(grid!!.maxY, robot!!.position.y)
    }

    @Test
    fun testExecuteCommand_SouthBlocked() {
        robot!!.position.y = 0
        robot!!.position.direction = Position.Direction.S
        robot!!.executeCommand('M')
        Assertions.assertEquals(0, robot!!.position.y)
    }

    @Test
    fun testExecuteCommand_WestBlocked() {
        robot!!.position.x = 0
        robot!!.position.direction = Position.Direction.W
        robot!!.executeCommand('M')
        Assertions.assertEquals(0, robot!!.position.x)
    }
}
